python/ray/experimental/channel/hccl_group.py

```diff
@@ -0,0 +1,168 @@
+            comm_id: A unique communicator ID.
+            rank: The rank of this actor. If None, then the caller is not a
+                participant of the HCCL group.
+            actor_handles: A list of actor handles, in rank order.
+            device_id: The NPU device id to use for HCCL operations.
+        """
+        self._world_size = world_size
+        self._rank: Optional[int] = rank
+        self._actor_handles = actor_handles
+        self._device_id = device_id
+
+        if rank is not None:
+            assert ray.get_gpu_ids(), "HCCL actor has no NPUs assigned"
+            assert device_id is not None, "HCCL actor must specify device_id"
+
+            expected_rank = self.get_rank(ray.get_runtime_context().current_actor)
+            assert (
+                rank == expected_rank
+            ), f"HCCL actor's rank {rank} does not match expected rank {expected_rank}"
+
+            import torch
+            import torch_npu
+            import torch.distributed as dist
+
+            # Initialize HCCL process group
+            os.environ['MASTER_ADDR'] = '127.0.0.1'
+            os.environ['MASTER_PORT'] = '29500'
+            os.environ['HCCL_WHITELIST_DISABLE'] = '1'
+            torch_npu.npu.set_device(device_id)
+            dist.init_process_group(backend='hccl', world_size=world_size, rank=rank)
+
+            self._comm = dist
+        else:
+            self._comm = None
+
+        self._closed = False
+
+    def initialize(self, rank: int) -> None:
+        # No additional initialization is needed.
+        pass
+
+    def get_actor_handles(self) -> List["ray.actor.ActorHandle"]:
+        return self._actor_handles
+
+    def get_rank(self, actor: ray.actor.ActorHandle) -> int:
+        """
+        Return the given actor's rank in the HCCL communicator.
+
+        Args:
+            actor: The actor handle to look up.
+        """
+        actor_ids = [a._ray_actor_id for a in self._actor_handles]
+        try:
+            rank = actor_ids.index(actor._ray_actor_id)
+        except ValueError:
+            raise ValueError("Actor is not in the HCCL group.")
+        return rank
+
+    def get_self_rank(self) -> Optional[int]:
+        """
+        Return this actor's rank.
+        """
+        return self._rank
+
+    def get_world_size(self) -> int:
+        """
+        Return the number of ranks in the HCCL communicator.
+        """
+        return self._world_size
+
+    def send(self, value: "torch.Tensor", peer_rank: int) -> None:
+        """
+        Send a torch.Tensor to a peer.
+
+        Args:
+            value: The torch.Tensor to send. It should already be on this
+                actor's NPU device.
+            peer_rank: The rank of the actor to send to.
+        """
+        if self._closed:
+            raise RayChannelError("HCCL group has been destroyed.")
+        
+        self._comm.send(tensor=value, dst=peer_rank)
```

Comment:
I think collective is a more general module that can be used for all other ray module. While here we need a module specify for aDAG channel. I think we can try to have another PR for hccl_collective_group so it can be used as a utils so we can use the NPU easier. In collective we can try to solve the double import or other problems that we meet.

Reply:
So in yesterday's aDAG meeting someone mentioned `nccl_collective_group` is actually old code, and `nccl_group` send/receive is what's currently used. We can discuss more to see how to extend it to support collectives it as apart of the refactor proposal.

Reply:
There is another PR to support collective fn as a node type https://github.com/ray-project/ray/pull/47621

I see they implemented `[collective/allreduce.py](https://github.com/ray-project/ray/pull/47621/files#diff-131c20c8eb3ed5eda3814e55f14abd73f1c9f444da42deeb89d782fbcfc00e13)` which calls `allreduce` of the `GPUCommunicator` in `nccl_group.py`